File: go/sentinel/get_indexes_from_image.go
package sentinel

import (
	"math"

	"github.com/lukeroth/gdal"
)

func GetIndexesFromImage(dataset gdal.Dataset) map[string][][]float64 {
	// Read bands
	bands := map[string]gdal.RasterBand{
		"B05": dataset.RasterBand(1),
		"B08": dataset.RasterBand(2),
		"B11": dataset.RasterBand(3),
		"B02": dataset.RasterBand(4),
		"B04": dataset.RasterBand(5),
		"B06": dataset.RasterBand(6),
		"CLD": dataset.RasterBand(7),
		"SCL": dataset.RasterBand(8),
	}

	// Read data from bands
	readBand := func(band gdal.RasterBand) [][]float64 {
		xSize := band.XSize()
		ySize := band.YSize()
		data := make([]float64, xSize*ySize)
		band.IO(gdal.RWFlag(gdal.Read), 0, 0, xSize, ySize, data, xSize, ySize, 0, 0)
		result := make([][]float64, ySize)
		for i := range result {
			result[i] = data[i*xSize : (i+1)*xSize]
		}
		return result
	}

	bandData := make(map[string][][]float64)
	for key, band := range bands {
		bandData[key] = readBand(band)
	}

	// Calculate indexes
	ndre := calculateIndex(bandData["B08"], bandData["B05"])
	ndmi := calculateIndex(bandData["B08"], bandData["B11"])
	psri := calculateIndex(bandData["B04"], bandData["B06"])
	ndvi := calculateIndex(bandData["B08"], bandData["B04"])

	indexes := map[string][][]float64{
		"ndre":  ndre,
		"ndmi":  ndmi,
		"psri":  psri,
		"ndvi":  ndvi,
		"b02":   bandData["B02"],
		"b04":   bandData["B04"],
		"cloud": bandData["CLD"],
		"scl":   bandData["SCL"],
	}

	return indexes
}

func calculateIndex(band1, band2 [][]float64) [][]float64 {
	rows := len(band1)
	cols := len(band1[0])
	result := make([][]float64, rows)
	for i := range result {
		result[i] = make([]float64, cols)
		for j := range result[i] {
			denominator := band1[i][j] + band2[i][j]
			if denominator != 0 {
				result[i][j] = (band1[i][j] - band2[i][j]) / denominator
			} else {
				result[i][j] = 0
			}
		}
	}
	return result
}

func validateIndexes(psri, ndvi, ndmi, ndre, cld, scl, b02, b04 float64, weather map[string]interface{}) bool {
	if math.IsNaN(psri) || math.IsNaN(ndvi) || math.IsNaN(ndmi) || math.IsNaN(ndre) {
		return false
	}
	if cld > 0 || scl == 3 || scl == 8 || scl == 9 || scl == 10 {
		return false
	}
	if (b04+b02)/2 > 0.9 {
		return false
	}
	if weather["precipitation"] == nil || weather["temperature"] == nil {
		return false
	}
	if psri == 0 && ndvi == 0 && ndmi == 0 && ndre == 0 {
		return false
	}
	return true
}

File: python/get_indexes_from_image.py
import rasterio
import numpy as np
from io import BytesIO

def get_indexes_from_image(image):
    if isinstance(image, BytesIO):
        dataset = rasterio.open(image)
    else:
        dataset = rasterio.open(BytesIO(image))
    
    with dataset:
        bands = {
            "B05": dataset.read(1),  # Band 1: B05
            "B08": dataset.read(2),  # Band 2: B08
            "B11": dataset.read(3),  # Band 3: B11
            "B02": dataset.read(4),  # Band 4: B02
            "B04": dataset.read(5),  # Band 5: B04
            "B06": dataset.read(6),  # Band 6: B06
            "CLD": dataset.read(7),  # Band 7: Cloud probability (CLD)
            "SCL": dataset.read(8)   # Band 8: Scene classification (SCL)
        }
        
        # Calculate indexes
        ndre = np.divide((bands["B08"] - bands["B05"]), (bands["B08"] + bands["B05"]), out=np.zeros_like(bands["B08"]), where=(bands["B08"] + bands["B05"])!=0)
        ndmi = np.divide((bands["B08"] - bands["B11"]), (bands["B08"] + bands["B11"]), out=np.zeros_like(bands["B08"]), where=(bands["B08"] + bands["B11"])!=0)
        psri = np.divide((bands["B04"] - bands["B02"]), bands["B06"], out=np.zeros_like(bands["B04"]), where=bands["B06"]!=0)
        ndvi = np.divide((bands["B08"] - bands["B04"]), (bands["B08"] + bands["B04"]), out=np.zeros_like(bands["B08"]), where=(bands["B08"] + bands["B04"])!=0)
        
        indexes = {
            'ndre': ndre,
            'ndmi': ndmi,
            'psri': psri,
            'ndvi': ndvi,
            'b02': bands["B02"],
            'b04': bands["B04"],
            'cloud': bands["CLD"],
            'scl': bands["SCL"]
        }
    
    return indexes

def validate_indexes(psri_value, ndvi_value, ndmi_value, ndre_value, cld_value, scl_value, b02_value, b04_value, weather):
    return  not (np.isnan(psri_value) or np.isnan(ndvi_value) or np.isnan(ndmi_value) or np.isnan(ndre_value) or cld_value > 0 or scl_value in [3, 8, 9, 10] or (b04_value + b02_value) / 2 > 0.9 or weather['precipitation'] is None or weather['temperature'] is None or (psri_value == 0 and ndvi_value == 0 and  ndmi_value == 0 and ndre_value == 0))